tools/eliminator/node_modules/uglify-js/lib/process.js

```diff
@@ -1681,7 +1681,7 @@ function gen_code(ast, options) {
                         }
                         if (member(rvalue[0], [ "assign", "conditional", "seq" ]) ||
                             rvalue[0] == "binary" && PRECEDENCE[operator] >= PRECEDENCE[rvalue[1]] &&
-                            !(rvalue[1] == operator && member(operator, [ "&&", "||", "*" ]))) {
```

Comment:
Should this patch go to upstream UglifyJS? https://github.com/mishoo/UglifyJS. Seems critical if so.

Also this might be something to be treated like `-ffast-math` in native compilers? In `-ffast-math` mode, compiler gets a hint that it can treat floating point math with associativity, and reorder in ways that are not strictly identical in floating point math. Not all developers need to care about this, so they would be taking a possibly needless optimization penalty here?

Reply:
OK, I'll try to submit a PR to UglifyJS later.

As to  `-ffast-math` - can the extra parentheses really result in an optimization penalty?

Reply:
> aggressive variable elimination would produce something like "result = huge_num * (huge_num * almost_zero)", which would be turned by uglify into "result = huge_num * huge_num * almost_zero", which would produce infinity.

> As to `-ffast-math` - can the extra parentheses really result in an optimization penalty?

Hmm, it is a bit involved I guess. Say, if you had `r = a*(b*(c*d));`, then under `-ffast-math`, you can reorder that to `r = (a*b)*(c*d);` to compute in parallel. I write it in parentheses like that, since if a JS engine saw just `r = a*b*c*d;` it would never take the liberty of computing `(a*b)*(c*d)`, so with manually choosing parentheses, one would coerce the JS engine to perform the fast math style result.

However floating point math is commutative, so `a*b=b*a` and `a+b=b+a`, so given `r = a*(b*c)` a tight optimization would be to turn it to `r=b*c*a` (if `a`,`b` and `c` aren't functions with side effects), that would retain proper computation order, while generating minimal code size.

Converting `r = a*(b*c)` to `r = a*b*c` could perhaps be seen as a sloppy `-ffast-math`-like behavior to save two bytes from the generated output (unless uglify does something more clever in more complex scenarios). But if we don't consider `-ffast-math`, the transformation `r = a*(b*c)` to `r = a*b*c` is certainly a bug.

Would be good to post that to upstream, and if they are green to land it, can then apply the same patch directly here if you don't want to do the hassle of full UglifyJS version upgrade. (whichever you prefer - unless @kripken has a thought here). If hot-patching UglifyJS, please add a comment `/* XXX Emscripten: backporting fix from https github com / PR-number */` or something like that to leave a trail for the next person who is looking at upgrading Uglify versions.

Reply:
I think Uglify1 is no longer being developed, so possibly they wouldn't be interested in this, but yeah, can't hurt to open a PR there. There's no risk of us needing to upgrade Uglify1, though, I wouldn't worry about that.

I would add an `// XXX EMSCRIPTEN - removed "*" since it is not associative` comment here, which we've done for our local changes.

Other than that, looks good to me, although probably we should measure the impact on asm.js code size - probably not significant, but just to be sure.

Reply:
FYI: I created a PR for Uglify 1 (https://github.com/mishoo/UglifyJS/pull/525) but haven't got a reply so far.